### PR TITLE
fix: fix #13684

### LINKED ISSUE #13684: java.lang.IllegalArgumentException: Metrics port cannot be null
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version: 3.2.9
* Operating System version: CentOS 8
* Java version: Oracle JDK 8u221

### Steps to reproduce this issue

1. set system property `dubbo.metrics.enabled=false`
2. define dubbo references by exporting `ReferenceConfig.get()` as `Bean`s 
3. start

Pls. provide [GitHub address] to reproduce this issue.

### Expected Behavior

Project starts successfully. References created without metrics configuration.

### Actual Behavior

References created with metrics enabled, but port is null.

Exception trace:

```
2024-01-22 17:55:47.207 |-ERROR [main] org.apache.dubbo.config.deploy.DefaultApplicationDeployer []-[traceId:] -|  [DUBBO] Dubbo Application[1.1](qijian-rec-service-41-prod) an exception occurred when handle starting event, dubbo version: 3.2.9, current host: 10.5.139.33, error code: 5-14. This may be caused by , go to https://dubbo.apache.org/faq/5/14 to find instructions. 
java.lang.IllegalArgumentException: Metrics port cannot be null
	at org.apache.dubbo.common.utils.Assert.notEmptyString(Assert.java:33) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.AbstractInterfaceConfig.appendMetricsCompatible(AbstractInterfaceConfig.java:280) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ServiceConfig.buildAttributes(ServiceConfig.java:662) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ServiceConfig.doExportUrlsFor1Protocol(ServiceConfig.java:587) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ServiceConfig.doExportUrls(ServiceConfig.java:579) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ServiceConfig.doExport(ServiceConfig.java:540) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ServiceConfig.export(ServiceConfig.java:340) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ServiceConfigBase.export(ServiceConfigBase.java:422) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.metadata.ConfigurableMetadataServiceExporter.export(ConfigurableMetadataServiceExporter.java:76) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.metadata.ExporterDeployListener.onModuleStarted(ExporterDeployListener.java:85) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultApplicationDeployer.exportMetadataService(DefaultApplicationDeployer.java:1263) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultApplicationDeployer.prepareApplicationInstance(DefaultApplicationDeployer.java:764) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultApplicationDeployer.checkState(DefaultApplicationDeployer.java:1149) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultApplicationDeployer.notifyModuleChanged(DefaultApplicationDeployer.java:1137) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultModuleDeployer.onModuleStarted(DefaultModuleDeployer.java:363) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultModuleDeployer.startSync(DefaultModuleDeployer.java:188) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.deploy.DefaultModuleDeployer.start(DefaultModuleDeployer.java:156) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ReferenceConfig.get(ReferenceConfig.java:235) ~[dubbo-3.2.9.jar:3.2.9]
	at org.apache.dubbo.config.ReferenceConfigBase.get(ReferenceConfigBase.java:395) ~[dubbo-3.2.9.jar:3.2.9]
```

### Analysis

In https://github.com/apache/dubbo/blob/2cf54f7947596abbf92f907853a2df7cbfc8b7fb/dubbo-common/src/main/java/org/apache/dubbo/config/AbstractInterfaceConfig.java#L274-L277 (though it is marked as deprecated but is still used now), `MetricsConfig` is got by `getConfigManager().getMetrics().orElse(null)`. Dive in we can know that the instance actually comes from https://github.com/apache/dubbo/blob/2cf54f7947596abbf92f907853a2df7cbfc8b7fb/dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java#L294-L296. It is added here https://github.com/apache/dubbo/blob/2cf54f7947596abbf92f907853a2df7cbfc8b7fb/dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java#L554-L566 because the statement `ConfigurationUtils.hasSubProperties(configurationMaps, AbstractConfig.getTypePrefix(cls))` is `true` here, which I think it should do more than just looking for properties with `dubbo.metrics` prefix.

### Workaround

Now we work it around by adding `org.apache.dubbo:dubbo-spring-boot-observability-starter` to avoid null port exceptions.

### Solution

One reasonable solution I think is to add one more condition after https://github.com/apache/dubbo/blob/2cf54f7947596abbf92f907853a2df7cbfc8b7fb/dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java#L554 to check if contains `${prefix}.enabled=false` configuration. Or maybe some solutions are on the way as `appendMetricsCompatible` is deprecated.

### PR DESCRIPTION
**Pull request by yuluo-yx**
## What is the purpose of the change

- Fix #13684 .

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java [hunk 1]
```diff
@@ -551,7 +551,18 @@ public <T extends AbstractConfig> List<T> loadConfigsOfTypeFromProps(Class<T> cl
...
             List<Map<String, String>> configurationMaps = environment.getConfigurationMaps();
-            if (ConfigurationUtils.hasSubProperties(configurationMaps, AbstractConfig.getTypePrefix(cls))) {
+
+            // check type prefix.
+            String typePrefix = AbstractConfig.getTypePrefix(cls);
+            boolean hasSubProperties = ConfigurationUtils.hasSubProperties(configurationMaps, typePrefix);
+
+            // check ${prefix}.enabled property configuration.
+            boolean metricsEnabled = configurationMaps.stream().anyMatch(configurationMap -> {
+                String key = typePrefix + ".enabled";
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java]: Should not hard code here

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java (hunk 1 above)
Comment (yuluo-yx) [reply dubbo-common/src/main/java/org/apache/dubbo/config/context/AbstractConfigManager.java]: I updated the code, extracted it into the commonconstant definition. pls review again. @AlbumenJ

**Comment by bitgorust** [comment]
@AlbumenJ @yuluo-yx As discussed with @songxiaosheng in https://github.com/apache/dubbo/issues/13684, a new pull request https://github.com/apache/dubbo/pull/13705 is submitted to solve this problem in another way.

**Comment by AlbumenJ** [comment]
Fixed in https://github.com/apache/dubbo/pull/13717

**Pull request closed without merging**